### PR TITLE
[CIS-1705] Run UI tests on CI

### PR DESCRIPTION
**Pull request by alteral**
### 🔗 Issue Link
- https://stream-io.atlassian.net/browse/CIS-1705

### 🎯 Goal

- We are going to run UI tests on CI (on release, push and PRs)

### 🛠 Implementation

- Integrate UI tests fastlane lane into `smoke-checks` workflow on Github Actions
- Implement a retry strategy to look after the tests for the first time

### 📝 Changes

- Create `xctestplan` for UI tests with mock server
- Stabilize the mock server and update the tests to avoid flakiness
- Implement user objects into the mock server
- Fetch `StreamChatTestHelpers` from `develop` until active development is finished
- Fetch `Swifter` from `stable`, it works faster and looks a little more stable at first glance 

### 🎨 UI Changes

Screen Shot 2022-03-31 at 3 25 40 PM

### 🧪 Testing

```bash
bundle exec fastlane test_e2e_mock device:"iPhone 11"
```

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] This change follows zero ⚠️ policy (required)
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎉 GIF

gif

### CODE REVIEW

File: .github/workflows/smoke-checks.yml [hunk 1]
```diff
@@ -58,6 +58,8 @@ jobs:
     runs-on: macos-11
     steps:
     - uses: actions/checkout@v2
+      with:
+        fetch-depth: 100
```
Comment (bielikb) [inline .github/workflows/smoke-checks.yml:62]: Good one. My hopes here are we hopefully wont have PRs with branches containing that many commits :)

File: Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload.swift [hunk 2]
```diff
@@ -28,7 +28,7 @@ struct ChannelPayload: Decodable {
     
     let isHidden: Bool?
 
-    private enum CodingKeys: String, CodingKey {
+    internal enum CodingKeys: String, CodingKey {
```
[the remark this answers is not in the document]
Comment (bielikb) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload.swift:31]:
```suggestion
    enum CodingKeys: String, CodingKey {
```

File: StreamChatUITestsApp/StreamChat/StreamChatWrapper.swift [hunk 3]
```diff
@@ -34,8 +35,48 @@ final class StreamChatWrapper {
...
         let query = ChannelListQuery(filter: .containMembers(userIds: [userCredentials?.id ?? ""]))
         let controller = ChatClient.shared.channelListController(query: query)
-        let channelList = ChatChannelListVC.make(with: controller)
+        let channelList = ChannelList.make(with: controller)
         return channelList
     }
 
 }
+
+final class ChannelList: ChatChannelListVC, ChatConnectionControllerDelegate {
```
Comment (bielikb) [inline StreamChatUITestsApp/StreamChat/StreamChatWrapper.swift:44]: @alteral please move this class to separate file.

File: StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift [hunk 4]
```diff
@@ -10,10 +10,19 @@ extension StreamMockServer {
     // TODO: CIS-1686
     func configureChannelEndpoints() {
         server[MockEndpoint.query] = { _ in
-            .ok(.text(TestData.getMockResponse(fromFile: .httpChannel)))
+            self.mockChannel(mockFile: .httpChannels)
```
Comment (bielikb) [inline StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift:13]: `weak`ify self here.

File: StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift [hunk 5]
```diff
@@ -10,10 +10,19 @@ extension StreamMockServer {
...
         }
         server[MockEndpoint.channels] = { _ in
-            .ok(.text(TestData.getMockResponse(fromFile: .httpChannelsQuery)))
+            self.mockChannel(mockFile: .httpChannelQuery)
```
Comment (bielikb) [inline StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift:16]: `weak`ify self here.

File: .github/workflows/smoke-checks.yml (hunk 1 above)
Comment (alteral) [reply .github/workflows/smoke-checks.yml:62]: 🤞

File: Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (alteral) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload.swift]: @bielikb, you sure? There are quite a lot of places where it's used. Please review this [commit](https://github.com/GetStream/stream-chat-swift/pull/1888/commits/cc845e68663b9f02c112e8ff108ce2b5e10e27ce).

File: StreamChatUITestsApp/StreamChat/StreamChatWrapper.swift (hunk 3 above)
Comment (alteral) [reply StreamChatUITestsApp/StreamChat/StreamChatWrapper.swift:44]: 👍

File: StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift (hunk 5 above)
Comment (alteral) [reply StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift:16]: If it's weak, how could I unwrap it?

File: StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift (hunk 4 above)
Comment (alteral) [reply StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift:13]: Do you think it's required? If it's weak, how could we unwrap it?

File: Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (bielikb) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/ChannelListPayload.swift]: I see, for all the new code written we follow the `non-internal` suggestion. Also the upcoming work related to `swiftlint` will cover the existing cases, let's not bring more divergence :)

https://github.com/GetStream/stream-chat-swift/pull/1866#discussion_r834156302

File: StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift (hunk 4 above)
Comment (bielikb) [reply StreamChatUITestsAppUITests/MockServer/ChannelResponses.swift:13]: @alteral you could return `empty` value
```
self?.mockChannel(mockFile: .httpChannels) ?? .badRequest(nil)
```

or leverage other strategies when weakifying `self`.

Lets avoid memory issues.

**Review by bielikb — APPROVED** [review]

**Pull request merged**